vavr/src/main/java/io/vavr/collection/Iterator.java

```diff
@@ -2042,281 +2156,386 @@ public T getNext() {
     }
 
-    final class CachedIterator<T> extends AbstractIterator<T> {
+    private static Object[] take(Iterator<?> source, Object[] target, int offset, int count) {
+        int i = offset;
+        while (i < count + offset && source.hasNext()) {
+            target[i] = source.next();
+            i++;
+        }
+        if (i < target.length) {
+            final Object[] result = new Object[i];
+            System.arraycopy(target, 0, result, 0, i);
+            return result;
+        } else {
+            return target;
+        }
+    }
+}
 
-        private final Iterator<T> that;
+final class SingletonIterator<T> implements Iterator<T> {
 
-        private T next;
-        private boolean cached = false;
+    private final T element;
+    private boolean hasNext = true;
 
-        CachedIterator(Iterator<T> that) {
-            this.that = that;
-        }
+    SingletonIterator(T element) {
+        this.element = element;
+    }
 
-        @Override
-        public boolean hasNext() {
-            return cached || that.hasNext();
-        }
+    @Override
+    public boolean hasNext() {
+        return hasNext;
+    }
 
-        @Override
-        public T getNext() {
-            if (cached) {
-                T result = next;
-                next = null;
-                cached = false;
-                return result;
-            } else {
-                return that.next();
-            }
+    @Override
+    public T next() {
+        if (!hasNext()) {
+            throw new NoSuchElementException();
         }
+        hasNext = false;
+        return element;
+    }
 
-        T touch() {
-            next = next();
-            cached = true;
-            return next;
+    @Override
+    public void forEachRemaining(Consumer<? super T> action) {
+        Objects.requireNonNull(action);
+        if (hasNext) {
+            action.accept(element);
+            hasNext = false;
         }
     }
 
-    final class BigDecimalHelper {
+    @Override
+    public String toString() {
+        return "SingletonIterator";
+    }
+}
 
-        @GwtIncompatible("Math::nextDown is not implemented")
-        private static final Lazy<BigDecimal> INFINITY_DISTANCE = Lazy.of(() -> {
-            final BigDecimal two = BigDecimal.valueOf(2);
-            final BigDecimal supremum = BigDecimal.valueOf(Math.nextDown(Double.POSITIVE_INFINITY));
-            BigDecimal lowerBound = supremum;
-            BigDecimal upperBound = two.pow(Double.MAX_EXPONENT + 1);
-            while (true) {
-                final BigDecimal magicValue = lowerBound.add(upperBound).divide(two, HALF_UP);
-                if (Double.isInfinite(magicValue.doubleValue())) {
-                    if (areEqual(magicValue, upperBound)) {
-                        return magicValue.subtract(supremum);
-                    }
-                    upperBound = magicValue;
-                } else {
-                    lowerBound = magicValue;
+final class BigDecimalHelper {
```

Comment:
I would like to get rid of this class. Need to take a look at rangeBy(double, double) later...
But I think that it isn't possible to fix floating point arithmetic inaccuracy another way.